### DISCUSSION TITLE
Best way to use macOS specific API in 2022

### DISCUSSION THREAD

**Discussion by veler** [AvaloniaUI/Avalonia#8990] [Q&A] [answered]
Hello,

I'm brand new at Avalonia and I'm working on my first cross-platform project. I'm trying to do something I usually do easily in WPF or UWP: offer the user to choose the app theme between:
- Light
- Dark
- Use system theme << this one should listen to the system for changes on the theme and switch automatically between Light and Dark.

I know how to do this last option for Windows. I believe I found how to do it with Swift on macOS (https://github.com/bouk/dark-mode-notify). But I'm scratching my head around on how to do it with Avalonia.

So far, I understand the following (correct me if it's wrong, please):
* `MonoMac` should allow me to do that. But it's archived for many years so it might not be the ideal NuGet to use.
* `Xamarin.Macios` should also allow me to do that, but it sounds like it's not really compatible with Avalonia?
* It sounds like `Avalonia.Native` will not allow me to do that, or if it can, I don't figure out how.

In the end, I'm confused at:
* What's the difference between MonoMac, Xamarin.Macios, Avalonia.Native?
* Does Avalonia.Native miss things that MonoMac can do?
* Which one I should use for a new project? Should I consider using a combination of them (MonoMac along with Avalonia.Native for example)?
* Whatever option I use, if there's no wrapper around the macOS API I need to use, what's the process to do create an interop between SWIFT and .Net 6.0? Is there a way to ship build a SWIFT library class and do a `[DllImport]` or something like this? Should I consider running some `bash` instead? Anything else?

Thank you very much in advance for your guidance :)

**Comment by maxkatz6** [comment, marked as answer]
I will try to help here:

- MonoMac should allow me to do that. But it's archived for many years so it might not be the ideal NuGet to use.

Yes. You can find still working MonoMac.NetStandard package. But it doesn't have new APIs and won't work well on ARM. 

- Xamarin.Macios should also allow me to do that, but it sounds like it's not really compatible with Avalonia?

It is compatible, but you should use .NET 6 macos target version. I.e. `<TargetFramework>net6.0-macos. `

- It sounds like Avalonia.Native will not allow me to do that, or if it can, I don't figure out how

Well, it was created for internal avalonia use before net6-mac was released. And even now it gives a bit more flexibility for us. You can add new custom functiona there, compile as your native library and ship with your app. It can be accessed with MicroCOM. While I know some projects do it in this way, it might be not the easiest one. 

Yet another option - write your own native library and pinvoke it. 

My personal choice for my projects - net6-macos target framework.

**Reply by maxkatz6** [reply]
- What's the difference between MonoMac, Xamarin.Macios, Avalonia.Native?

Xamarin replaced MonoMac, but was tied to Mono framework for a long time. Now with net6 it works with NET Core. These APIs map macOS apis as close as possible with PInvokes. 
Avalonia.Native was created before net6 and works using custom COM interop. COM object model is exposed from the native library and glued to c# with code generator. 

- Does Avalonia.Native miss things that MonoMac can do?

Avalonia.Native has implemented only these functions what are needed for the Avalonia. Nothing more. 

- Which one I should use for a new project? Should I consider using a combination of them (MonoMac along with Avalonia.Native for example)?
Whatever option I use, if there's no wrapper around the macOS API I need to use, what's the process to do create an interop between SWIFT and .Net 6.0? Is there a way to ship build a SWIFT library class and do a [DllImport] or something like this? Should I consider running some bash instead? Anything else?

Unfortunately I am not sure if interop with Swift exists. But if there is alternative API for ObjC, there shouldn't be any blockers.

**Reply by maxkatz6** [reply]
And finally, yet another alternative is to PInvoke native API directly like here. But it would get more complicated if you need to pass a object or event handler (basically reimplementing monomac for that). 
https://github.com/AvaloniaUI/Avalonia/pull/8880/files

**Reply by maxkatz6** [reply]
So, if you want to go easiest and most pro-MS way, you should take a look at .NET 6 macos support.

**Reply by veler** [reply]
Thank you for your detailed and quick answer.
Since I'm using .net6, using net6.0-macos with the MacOs specific namespace is the obvious answer. I don't why I didn't think about it initially.
Thank you so much for your help! 😁
I will dig out how to call these apis using Net 6 MacOs.
